### ISSUE TITLE
[java] False-Positive: UnnecessaryImport in test code

### ISSUE THREAD

**Issue by zodac** [pmd/pmd#4712]
**Affects PMD Version:** 7.0.0-rc4

**Rule:** [UnnecessaryImport](https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#unnecessaryimport)

**Description:** I am getting false-positives for imports that are being flagged as not used in my test code. This is impacting both static and non-static imports. I haven't see this occur in any other code, only my test classes.

Note, that this seems to occur intermittently. If I execute on just the module containing this code it fails, but if I run on the whole project it passed fine. I am using **mvn clean install**, then using a profile that executes PMD. However, I see in Github actions that the failures occur there even when running the whole project:
https://github.com/zodac/folding-stats/actions/runs/6523823534/job/17714841008

**Code Sample demonstrating the issue:**
https://github.com/zodac/folding-stats/blob/master/testsuite-integration/src/test/java/me/zodac/folding/test/integration/TeamCompetitionStatsTest.java

**Expected outcome:**

PMD reports a violation at lines 22-40, and also L49, (all of which are imports from my own project, though other classes from my project are not reported as an issue), but they are all being used in the class.

**Running PMD through:** Maven (3.21.1-pmd-7.0.0-SNAPSHOT)

**Comment by tombriggsallego** [comment]
I too am seeing tons of false positives for this check with 7.0.0-rc4. I've been running PMD against one of our projects for more than five years years. With 6.55 I get a clean report, but today I tried upgrading to 7.0.0-rc4 and I get a ton of these false positives. I'm running pmd through ant but as noted I've been doing so successfully for quite a while so I doubt that's the issue.

**Comment by tombriggsallego** [comment]
Ok so I did some debugging on this and, at least for me, the problem stems from this:

```
        if (node.getQualifier() == null) {
            OverloadSelectionResult overload = node.getOverloadSelectionInfo();
            if (overload.isFailed()) {
                return null; // todo we're erring towards FPs
            }
```
which is in UnnecessaryImport.visit(ASTMethodCall node, Object data)

Our code has a bunch of occurrences of method calls that look like this:

```
            someMethod(argOne, argTwo,
                    new Arg3(SomeEnum.SOME_ENUM_VALUE, 0, ...));
```
That leads to two "unnecessary import" warnings - one for the SomeEnum class and one for the Arg3 class.

If I change the code to

```
           x = new Arg3(SomeEnum.SOME_ENUM_VALUE, 0, ...);
            someMethod(argOne, argTwo, x);
```
both errors go away. But I don't want to change a bunch of our code to quiet warnings. ;)

**Comment by oowekyala** [comment]
Interesting, thanks for looking into this

**Comment by UncleOwen** [comment]
Unfortunately, all the links in the OP are 404. Is there a way to reliably reproduce this?

**Comment by adangel** [comment]
The file has been moved to a different package: https://github.com/zodac-personal/folding-stats/blob/master/testsuite-integration/src/test/java/net/zodac/folding/test/integration/TeamCompetitionStatsTest.java

I didn't try to reproduce this, though.

**Comment by adangel** [comment]
Well, at least this comment...
https://github.com/zodac-personal/folding-stats/blob/543ace7a851a45d4518e8cb4f508faf75420f4cd/pom.xml#L912-L913
> `                <!-- Manual execution: -->`
> `                <!-- mvn pmd:check -Dlint -->`

...is wrong - this is the recipe to reproduce this error: first run `mvn clean` - and then run `mvn pmd:check` - the project is not compiled and the auxclasspath is incomplete.
Correct would be probably `mvn test-compile pmd:check -Dskip-linters=false`

In the original description however, the OP says he uses `mvn clean install` to run pmd, but it is not clear, whether he uses it to run the whole project (which is no problem) or whether he uses it to do a single module check (where he sees the problem).

It shouldn't fail on CI though, as there `mvn clean install -Dall` is used (https://github.com/zodac-personal/folding-stats/blob/543ace7a851a45d4518e8cb4f508faf75420f4cd/.github/workflows/java.yml#L72)
